### PR TITLE
Dunstrc: Actually use it and be less disruptive on (un)install

### PR DESCRIPTION
**Pull request by WhitePeter**
This PR proposes a fix for several issues with dunstrc. For starters, it was installing to the wrong location ('/etc/dunst/dunstrc') instead of at least ('/etc/xdg/dunst/dunstrc'), which would be the right place, were it not a distro managed confdir. On top of that 'make install' would happily overwrite whatever file is there already, violating at least the Debian Policy of never silently overwriting user/admin configs. Locally installed software is expected to put its system-wide conffiles in '/usr/local/etc'. In this case the XDG_CONF_DIRS spec (XDG base directory spec) comes into play, so the right location is '/usr/local/etc/xdg/dunst/dunstrc'. With theses changes, installation of 'dunstrc' is made to that path.
Plus, now install won't overwrite an already existing dunstrc, i.e. on upgrade. And it will also not remove it on uninstall. I introduce a new make target called "purge", blatantly borrowing the name from Debian, think: 'apt-get purge' as opposed to 'apt-get remove'. Also the current behavior (before this PR) is to just 'rm -rf /etc/dunst', which does not account for files the admin might have in there besides dunstrc, i.e. backups of said file or maybe even a git repo for versioning. The latter seems to be all the rage today for tinkering with configs. *Edit: But then again, judging by the Makefile, dunstrc was considered to be documentation and the current documentation, which I intend to update as well (see below), seems to do the same: "Copy it and make it work", in short.*

But there is a catch: XDG_CONF_DIRS might not be set (i.e. on Ubuntu), so it defaults to '/etc/xdg', which in turn would lead dunstrc being picked up from there and not the local etc. To account for that the systemd user service and the dbus service set XDG_CONF_DIRS appropriately in their respective environments. Which only leaves the case of running /usr/local/bin/dunst directly. I have not yet found a way to deal with that. So the user has a choice: run it and use a possibly outdated conffile provided by the distro, which might contain obsolete settings. Or make sure that they set XDG_CONF_DIRS, ie: 'env XDG_CONF_DIRS=/usr/local/etc/xdg:$XDG_CONF_DIRS /usr/local/dunst'.

Now, this PR is not *quite* finished, but I figured I might as well get some feedback before I update the README and also on the TODO list: remove non-existent settings from dunstrc which dunst keeps complaining about, now that it actually reads it.

### CODE REVIEW

**Comment by fwSmit** [comment]
> I have not yet found a way to deal with that. So the user has a choice: run it and use a possibly outdated conffile provided by the distro

I haven't looked at the PR yet, so no comment on that. I just wanted to say that dunst controls where it looks for conf files. We can just decide to look in /etc/local as well. The only thing is that the system installed version might also look in /etc/local.

**Comment by WhitePeter** [comment]
> > I have not yet found a way to deal with that. So the user has a choice: run it and use a possibly outdated conffile provided by the distro
> 
> I haven't looked at the PR yet, so no comment on that. I just wanted to say that dunst controls where it looks for conf files. We can just decide to look in /etc/local as well. The only thing is that the system installed version might also look in /etc/local.

/etc/local? Is that even a documented location or just an example to make the point that it would be possible?
And on the note about dunst deciding where to look, I was under the impression that this gets delegated to some external facility ATM, which would explain, that my changed version (build from 1875562705002fa3b0eec912bfa407d344060695), just picks it up when I point XDG_DATA_DIRS to /usr/local/etc/xdg an the file at dunst/dunstrc relative to that. Is there just some wrong default in the code? Because, as I mentioned before the XDG_DATA_DIRS is not explicitly set on Ubuntu, but I half expect to be the same on many other distros.

**Comment by WhitePeter** [comment]
Sorry, had to fix the Makefile, there was a conflict when I merged locally on while resolving it, I threw out the actual check it dunstrc already exist at SYSCONFFILE. I guess I was tired and got careless. :(

**Comment by fwSmit** [comment]
I like having the config file for system installs in /etc/dunst/dunstrc better than /etc/xdg/dunst/dunstrc. Take a look at #798 for when the config location was changed from /etc/dunst.

**Comment by fwSmit** [comment]
Take a look at src/settings.c for how the config location is currently decided

**Comment by WhitePeter** [comment]
> I like having the config file for system installs in /etc/dunst/dunstrc better than /etc/xdg/dunst/dunstrc. Take a look at #798 for when the config location was changed from /etc/dunst.

If *you* like your config in /etc as opposed to $(prefix)/etc, I believe you should use an override locally and not set this rather dangerous default. It violates the GNU make directories spec you referenced in #798. That one says:
>  This directory should normally be /usr/local/etc, but write it as **$(prefix)**/etc

Basically that is what I and probably most users want, the GNU defaults. At least that is what I expect when I run make without parameters, hence my initial report. ;) If one has good reason to and knows what they are doing and has accounted for the implications, then it is their prerogative to override those defaults, which is why they can be overridden. That is what package maintainer do routinely.

Also, your PR #798 names two different locations for dunstrc, the headline of the PR says /etc/dunst/dunstrc and the commit message says /etc/**xdg**/dunst/dunstrc. But ultimately, the code  uses /etc/dunst/dunstrc.

If I understand settings.c correctly, you only fallback on /etc/dunst/dunstrc when there is no other way to find it and a prior call to g_get_system_config_dirs() actually does return what XDG_CONF_DIRS says or its default, which is why the config in /etc/dunst/dunstrc has been a dud until now, at least when there is an actual dunstrc, provided by the distro package, in /etc/xdg/dunst. That is what I meant by external facility: [g_get_system_config_dirs()](https://valadoc.org/glib-2.0/GLib.Environment.get_system_config_dirs.html). Since you put dunstrc in a location that is not recognized by it, it will never return a path there and only when you have to fallback will the shipped dunstrc be picked up, while it _should_ be the first to be found since it is for a locally installed software "package" (the quotes because I like to distinguish it from distro packages).

As I pointed out, there is no default for XDG_CONF_DIRS, which actually points to /usr/local/etc/xdg, which is very unfortunate IMHO. But then again, I don't know all the reasoning that went into the spec. One thing immediately springs to mind: setting 'XDG_CONF_DIRS=/usr/local/etc/xdg:/etc/xdg' globally, i.e. in /etc/environment, as I have for now, should then lead to the _local_ etc always taking precedence, even for distro packages, if it works like PATH. That is also not really what I would expect. My thinking is: local software get its conf from /usr/local/etc, or /usr/local/etc/xdg in this case, and distro software uses /etc exclusively.

That's all for the time being. Now I need to check why the tests failed, again. :(

**Comment by WhitePeter** [comment]
BTW, any hints on how I debug those tests? The error messages don't tell me much, other than 'it was cancelled' or 'it failed'.

**Comment by WhitePeter** [comment]
Ah wait, I guess I figured it out. You just merged my other pull request and this one is still based on the previous state, i.e. the pkg-config checks which are now gone in the current master. Am I getting warmer? ;)

**Comment by WhitePeter** [comment]
OK, now the ubuntu-focal test fails because dunstrc was not removed on uninstall. But that is the expected behavior with this PR. ;)

**Comment by WhitePeter** [comment]
That did the trick. Yay! ;)

Please do note that this diverges from what the initial PR message suggests. For the time being it fixes the test runs, but I would like the new make target `uninstall-keepconf` to become the new default `uninstall` which is already only a pointer to the also new target `uninstall-purge` which is essentially only renamed from `prune`.

**Comment by fwSmit** [comment]
> /etc/local? Is that even a documented location or just an example to make the point that it would be possible?
> And on the note about dunst deciding where to look, I was under the impression that this gets delegated to some external facility ATM, which would explain, that my changed version (build from [2968d38](https://github.com/dunst-project/dunst/commit/2968d381c4c9718a8b60aa3145cac4ea6bb150cb)), just picks it up when I point XDG_DATA_DIRS to /usr/local/etc/xdg an the file at dunst/dunstrc relative to that. Is there just some wrong default in the code? Because, as I mentioned before the XDG_DATA_DIRS is not explicitly set on Ubuntu, but I half expect to be the same on many other distros.

To clarify, I meant `/usr/share/local/etc/`.

> 
> If _you_ like your config in /etc as opposed to $(prefix)/etc, I believe you should use an override locally and not set this rather dangerous default. It violates the GNU make directories spec you referenced in #798. That one says:
> 
> > This directory should normally be /usr/local/etc, but write it as **$(prefix)**/etc
> 
> Basically that is what I and probably most users want, the GNU defaults. At least that is what I expect when I run make without parameters, hence my initial report. ;) If one has good reason to and knows what they are doing and has accounted for the implications, then it is their prerogative to override those defaults, which is why they can be overridden. That is what package maintainers do routinely.

I mean that I don't like the xdg subdirectory in `${PREFIX}/etc`. I know that the standard that says to use the xdg subdirectory, but many programs seem to ignore it. I think it would be good to take a look at a few other programs to see what they're doing.

> Also, your PR #798 names two different locations for dunstrc, the headline of the PR says /etc/dunst/dunstrc and the commit message says /etc/**xdg**/dunst/dunstrc. But ultimately, the code uses /etc/dunst/dunstrc.

The `/etc/xdg/dunst/dunstrc` was a bad/outdated commit message that I noticed too late.

> If I understand settings.c correctly, you only fallback on /etc/dunst/dunstrc when there is no other way to find it and a prior call to g_get_system_config_dirs() actually does return what XDG_CONF_DIRS says or its default, which is why the config in /etc/dunst/dunstrc has been a dud until now, at least when there is an actual dunstrc, provided by the distro package, in /etc/xdg/dunst. That is what I meant by external facility: [g_get_system_config_dirs()](https://valadoc.org/glib-2.0/GLib.Environment.get_system_config_dirs.html). Since you put dunstrc in a location that is not recognized by it, it will never return a path there and only when you have to fallback will the shipped dunstrc be picked up, while it _should_ be the first to be found since it is for a locally installed software "package" (the quotes because I like to distinguish it from distro packages).
> 
> As I pointed out, there is no default for XDG_CONF_DIRS, which actually points to /usr/local/etc/xdg, which is very unfortunate IMHO. But then again, I don't know all the reasoning that went into the spec. One thing immediately springs to mind: setting 'XDG_CONF_DIRS=/usr/local/etc/xdg:/etc/xdg' globally, i.e. in /etc/environment, as I have for now, should then lead to the _local_ etc always taking precedence, even for distro packages, if it works like PATH. That is also not really what I would expect. My thinking is: local software get its conf from /usr/local/etc, or /usr/local/etc/xdg in this case, and distro software uses /etc exclusively. That is what I believe how it should be and does work with other software.
> 
> That's all for the time being. Now I need to check why the tests failed, again. :(
> 
> PS: Speaking of PATH I just realized that it could maybe leveraged to determine from which prefix dunst was started and just prepend that to what g_get_system_config_dirs() returns? Check argv[0], determine prefix and prepend to sysconfdir location if does not already point there.

`/usr/local/etc/dunst/dunst` seems like a good default to me. I haven't seen `/usr/local/etc/xdg` in a spec, only `/etc/xdg`,

**Comment by WhitePeter** [comment]
> I mean that I don't like the xdg subdirectory in `${PREFIX}/etc`. I know that the standard that says to use the xdg subdirectory, but many programs seem to ignore it. I think it would be good to take a look at a few other programs to see what they're doing.

But then the use `g_get_sysconfdirs()` is a pointless exercise. ;) The package maintainers use it, where appropriate. I believe this is only meant for desktop software (the D stands for Desktop), which is why shells, servers and whatnot do not use it. I actually like the approach of XDG, it got a lot of dot files in my home out of the way, but there are still more dot files than actual files and directories there, so they do have a point there. I believe the same thinking went into /etc/xdg. Since quite a lot of software comes with a whole bunch of conffiles they would make an unholy mess of things. At the time /etc was specified it was usually one rc file per program.

> `/usr/local/etc/dunst/dunst` seems like a good default to me. I haven't seen `/usr/local/etc/xdg` in a spec, only `/etc/xdg`,

From my POV `/usr/local/etc` should show the same behaviour as `/etc`, like it does for everything else. Since the default for XDG_CONF_DIRS is `/etc/xdg` it *should* be the same relative path with the `/usr/local` prefix. Otherwise you create a special case that needs to be handled. using `etc/xdg` only requires a change of PREFIX. While I do appreciate your preference I think you make it unnecessarily hard on yourself. I mean, look at the fallback you introduced in #798, talk about hardcoding. ;)

Again, if you personally prefer it that way, nothing is stopping you from achieving such a setup locally, that is why those are parameters that can be changed, but I strongly believe software should adhere to agreed upon standards where it can. It is all out there and very well documented. Please don't force your personal preferences on everybody else. Think about the principle of least surprise. We wouldn't be having this conversation if the current defaults hadn't surprised me. ;) Initially I only wanted to see if those pesky warnings about X(wayland) missing MIT_SCREEN_SAVER support would go away, which they did. I should not have to change the default parameters to just run `make && sudo make install`. Virtually every other software project I build locally works that way. Then, if one want to get fancy bells and whistles or some non-standard setup, one goes and digs into the parameters. Also think about less experienced users, that are having a hard time making sense of this procedure (`make and stuff`). They are happy when they get things running--I am also thinking about me 15-20 years ago. Those people might end up with a half broken system and don't have the capacity (yet) to know.

Look what happened in my setup: dunst 1.4.1 from the Ubuntu repo, then I build dunst from source, which on installation overwrites two system files and only because of your fallback preference did it not do the same with the system wide config. Now, I can deal with it, and please don't get me wrong, I am not complaining. I just want to make it better, so others, who might not even know it, don't experience the same. This could also save you from some peculiar issue reports in the long run: "Hey I have changed my system wide dunstrc, but dunst does not recognize the changes".

**Comment by WhitePeter** [comment]
This should solve all the issues discussed here. Use SYSCONF(DIR/FILE) to override the defaults at will. No point in getting fancy at runtime. As I said software in /usr/local has no business reading let alone writing in /etc. Look at other projects like mpv, or unbound. They do basically the same thing: string constant which gets set at configure/compile time. That way there can never arise the situation where it is unclear, which system conf to use. There should only be one. As @fwSmit said, dunst decides where, so make sure there is no ambiguity for no gain or reason.

**Comment by WhitePeter** [comment]
One more thing about 'to xdg or not xdg', just because you have not seen any software using it, does not mean there isn't any. Out ouf curiosity I ran this:
```sh
$ apt-file search /etc/xdg | cut -d' '  -f1 | sort | uniq | wc -l
288
```
And those are just the package names, that have at least one reference to said path. Since dunst is quite clearly a software that is bound to the desktop, etc/xdg is obviously the correct default. @bebehei seemed to think so too, apparently, when he introduced walking the XDG_CONFIG_DIRS by means of g_get_system_config_dirs() in 403e4cc176cde79f1ccabc30c878e5201e696d1b. But since it more likely than not, that this function returns /etc/xdg, it is just not suitable for the job, sadly.

**Comment by fwSmit** [comment]
> One more thing about 'to xdg or not to xdg', just because you have not seen any software using it, does not mean there isn't any. Out ouf curiosity I ran this:
> ```sh
> $ apt-file search /etc/xdg | cut -d' '  -f1 | sort | uniq | wc -l
> 288
> ```
> And those are just the package names, that have at least one reference to said path. Since dunst is quite clearly a software that is bound to the desktop, etc/xdg is obviously the correct default. @bebehei seemed to think so too, apparently, when he introduced walking the XDG_CONFIG_DIRS by means of g_get_system_config_dirs() in 403e4cc176cde79f1ccabc30c878e5201e696d1b. But since it more likely than not returns /etc/xdg, it is just not suitable for the job, sadly.
> 
> P.S.: Just noticed, that the decision was made way back by Michael Stapelberg in 78b6e2b19f1595b4c6ad38225f667a24d522a576, Benedikt just got rid of the dependency libxdg-basedir.

But is there also software using `/usr/local/etc/xdg`?

**Comment by fwSmit** [comment]
By the way, I do agree with your standpoint now. I'm just looking to implement it in the most user friendly way.

**Comment by WhitePeter** [comment]
> But is there also software using `/usr/local/etc/xdg` instead of `/usr/local/etc`? The reason I ask is that I didn't see anything mentioned about that in the specs

I don't see a reason to use a different sub directory just because the prefix changes. My understanding of /usr/local/etc is that it should be possible to just pack it up and move it to /etc with no other changes, i.e.:
```sh
tar -cC /usr/local etc | tar -xC /
```
So the xdg subdirectory should be in there as well. Plus, it is simplest solution, which does not need to care about the prefix.

About the specs (XDG base directories?): I believe, given my above expectations, the authors might not see a reason to be explicit about it, if /usr/local/etc is exactly like /etc, just with a shifted prefix.

> By the way, I do agree with your standpoint now. I'm just looking to implement it in the most user friendly way.

Great! :-) I think this is the most user friendly way--take it from a picky one. ;)

File: src/settings.c
Comment (WhitePeter) [reply src/settings.c:32]: > On the one hand you want Dunst to look at the conf dir you said to put the config file in, _but on the other hand it might confuse users that otherwise identical dunst installs behave differently._

Not this one right here. ;) That is exactly what I'd expect, because local software should stay out of everything that does not belong to its prefix. Plus, different behaviour should be expected anyway since one installs almost certainly a more up-to-date version from upstream, which could always involve deprecation of config settings. If users do want to share the config, they can just make $PREFIX/etc/xdg/dunst a symlink to /etc/xdg/dunst or the other way around.

As examples, again: mpv and unbound, which I happen to have installed locally and which make actual use of $PREFIX/etc. They even incorporate the prefix in the documentation.

> I like the env change better, since we can stick to a standard way to look up the config and users have the flexibility to decide what they want. 

But that is *less* user-friendly since it involves changing their environment, which I am not so sure anymore is/was the right thing to do in the first place. The way XDG_CONFIG_DIRS works and the current way dunst makes use of it, leads to situations where the user does not know, which config gets used. The default on Ubuntu is an unset XDG_CONFIG_DIRS, which in turn defaults to `/etc/xdg`, which is precisely what I *don't* want. Even if I set it up like this:
```sh
XDG_CONFIG_DIRS=/usr/local/etc/xdg:/etc/xdg
```
globally, i.e. in /etc/environment, it does not work as intended, because then any distro software which is 1. installed and 2. has also been installed locally, will then prefer /usr/local/etc, which is definitely wrong. There should be as much isolation between the two prefixes as possible, at least by default. Advanced users with special needs should not have a problem with making it work.
I also think now that g_get_system_config_dirs() has no use in regular software for precisely those reasons just outlined. It may have its use for programs that update those directories or whatever. If the authors intended it differently, well, they have left a huge gaping hole in the spec as to how to deal with locally installed software that does not lead to funny behaviour. ;) Maybe the intended use is to prepend PREFIX, when one queries XDG_CONFIG_DIRS. That could be a solution which would result in what I expect, but it is undocumented, and one should never rely on undocumented behaviour. And XDG_CONFIG_DIRS would then need to be queried at configure/compile time. Since it is not a simple string but an array split by ':', I would not want to deal with that if I can help it. I mean, checking it, possibly having to default, and if it has an actual value, splitting and walking it, seems like a lot of hassle, when the default on distros is unset anyway. I would sooner risk a user complaint about an issue in a rather distant future, and fix it then, than doing this stuff now, while the correct thing to do is still unclear. Anyone that runs a distro which actually does set this parameter, please come forward now. ;-)

> We should just instruct users to change their global environment when they want dunst to look at the config. For most users it won't be a problem anyways, since they make a user config that overrides the system one.

As I just outlined, that might be a bad idea, especially if they use a user config, because they might also install other software locally, keeping the distro version installed, not have a user config for it, and then all hell breaks loose: "Dunst broke my toys!!1!", or whatever, but not before they go running and complain to the distro maintainers. ;)
I think, for the time being, this solution is dead simple and should do the right thing in most cases. I can just make the previous code a comment, so it is still there, should it be necessary to reactivate it. But then, what is git for? ;)

**Comment by fwSmit** [comment]
> > But is there also software using `/usr/local/etc/xdg` instead of `/usr/local/etc`? The reason I ask is that I didn't see anything mentioned about that in the specs
> 
> I don't see a reason to use a different sub directory just because the prefix changes. My understanding of /usr/local/etc is that it should be possible to just pack it up and move it to /etc with no other changes, i.e.:
> 
> ```shell
> tar -cC /usr/local etc | tar -xC /
> ```
> 
> So the xdg subdirectory should be in there as well. Plus, it is simplest solution, which does not need to care about the prefix.
> 
> About the specs (XDG base directories?): I believe, given my above expectations, the authors might not see a reason to be explicit about it, if /usr/local/etc is exactly like /etc, just with a shifted prefix.

Okay, it's fine. Let's use `/usr/local/etc/xdg`. Sorry if I got too defensive

File: src/settings.c
Comment (fwSmit) [reply src/settings.c:32]: Sorry I forgot that we actually expect different behaviour from the two.

**Comment by WhitePeter** [comment]
> Okay, it's fine. Let's use `/usr/local/etc/xdg`. Sorry if I got too defensive

No problem. ;)

**Review by fwSmit — CHANGES_REQUESTED** [review]
I'll give you a go-ahead, but I suggested an easier way to implement what you intend. 
It's kind of unfortunate that we don't follow the XDG spec for system config dirs, but I don't see a better way either.

File: src/settings.c
Comment (WhitePeter) [reply src/settings.c:32]: Oh wait, I need to roll back the other changes which set XDG_CONFIG_DIRS. They should not hurt, since they don't do anything now, but they are unnecessary an might confuse people that take a look...

File: Makefile
Comment (fwSmit) [inline Makefile:65]: It seems like my comment didn't come through, so here is it again.
This can be implemented without requiring a separate file by using DEFAULT_CFLAGS in config.mk. See how it's done for wayland enabling/disabling.

File: src/settings.c
Comment (WhitePeter) [reply src/settings.c:32]: One more thing that did get pointed out to me on the XDG mailing list: The spec talks about "information" as opposed to files, so just because the one file in a more important location (i.e. the first) does not mean, you should not try to find *other* information in less important locations. So, technically dunst does not do that either, and if it would do that with i.e. `XDG_CONFIG_DIRS=/usr/local/etc/xdg:/etc/xdg`, then it can get really messy: Read setting A from /usr/local/etc/xdg... but since setting B is not there move on and read it from /etc/xdg... where it is set. That situation could also arise and would be the same for locally installed and distro software.

See the [spec](https://specifications.freedesktop.org/basedir-spec/latest/ar01s03.html) for reference:
> When the same **information** is defined in multiple places the information defined relative to the more important base directory takes precedent.

This means that the current way of dealing with configs is not technically correct, because it uses **either** the user config, if found, **or** the system-wide config otherwise. But this has nothing to do directly with this PR so I won't do anything about it. Just wanted to add this bit, since it bares *some* relevance.

File: Makefile
Comment (WhitePeter) [reply Makefile:65]: I'll have look and will get back at you...

**Comment by WhitePeter** [comment]
> It's kind of unfortunate that we don't follow the XDG spec for system config dirs, but I don't see a better way either.

But we are, the files go to the right places. Yes, XDG_CONFIG_DIRS does not get queried, but there is no point in doing so, at least for dunst since it only has one location and one file in that. Plus, it is the default location, so unless distro maintainers want to change it in their distro just for kicks, we should be fine. And if they do, they only need to set XDG_CONFIG_DIRS to also look in /etc/xdg, if some helper software needs to know that there are also config files in there.

File: src/settings.c
Comment (fwSmit) [reply src/settings.c:32]: Good to know. This isn't really high priority, since I want to keep the systemd wide dunstrc the same as the in-code defaults. So in practice it would only make a difference if one changes their system-wide config.
Just for reference, [this](https://lists.freedesktop.org/archives/xdg/2021-September/014560.html) is the mentioned email thread

Comment (fwSmit) [reply src/settings.c:32]: From the emails:

> HOWEVER, Base Directory does not specify a first match wins rule.  It
> attributes more importance (the spec's terminology) to files located
> in earlier directories in the list, but that does not imply that only
> one can be used.

From the spec:

> When the same information is defined in multiple places the information defined relative to the more important base directory takes precedent.

This doesn't sound like you *may not* use a first match wins rule.

Comment (WhitePeter) [reply src/settings.c:32]: > Just for reference, [this](https://lists.freedesktop.org/archives/xdg/2021-September/014560.html) is the mentioned email thread

Correct, should have linked it myself, sorry.

> This doesn't sound like you *may not* use a first match wins rule.

Yes, I think you are right, but it would be nice to have. Basically I think this could be done quite easily, though. If you look at shell initialization, for instance: Read the *least* important files first and work your way to the most important ones, or simply concatenate the files in that order and just overwrite values as they come. Since the last value comes from the most important file, that's it.

Haven't looked how much work that would take to change in dunst's code, though. I am holding off on that for now.

**Comment by WhitePeter** [comment]
Oh, I guess I shouldn't have squashed my last commits. Now I cannot find the conversation that went with it since git cannot find the commits anymore.

**Comment by WhitePeter** [comment]
A quick side not about config.mk/Makefile: shouldn't EXEC_PREFIX also be defined in there and subsequently BINDIR use that value? Like this:
```make
EXEC_PREFIX ?= ${PREFIX}
BINDIR ?= ${EXEC_PREFIX}/bin
```
I am asking because of this passage in the [GNU make spec](https://www.gnu.org/software/make/manual/make.html#Directory-Variables):
> **exec_prefix**
> A prefix used in constructing the default values of some of the variables listed below. The default value of exec_prefix should be $(prefix). (If you are using Autoconf, write it as ‘@exec_prefix@’.)
>
> Generally, $(exec_prefix) is used for directories that contain machine-specific files (such as executables and subroutine libraries), while $(prefix) is used directly for other directories.
>
> Running ‘make install’ with a different value of exec_prefix from the one used to build the program should not recompile the program.

I am not sure, but this could make live easier for package maintainers, given that they then can set PREFIX empty and EXEC_PREFIX=/usr. I mean, while I am at it... ;) Unless this goes to much out of scope for this PR, that is.

**Comment by fwSmit** [comment]
> Oh, I guess I shouldn't have squashed my last commits. Now I cannot find the conversation that went with it since git cannot find the commits anymore.

Sqaush should open an editor to merge the commit messages. Nothing gets lost by default.
Anyways, there are still 10 commits for a change that should probably be 1 commit, so please squash a bit more

**Comment by fwSmit** [comment]
> A quick side not about config.mk/Makefile: shouldn't EXEC_PREFIX also be defined in there and subsequently BINDIR use that value? Like this:
> ```make
> EXEC_PREFIX ?= ${PREFIX}
> BINDIR ?= ${EXEC_PREFIX}/bin
> ```
> I am asking because of this passage in the [GNU make spec](https://www.gnu.org/software/make/manual/make.html#Directory-Variables):
> > **exec_prefix**
> > A prefix used in constructing the default values of some of the variables listed below. The default value of exec_prefix should be $(prefix). (If you are using Autoconf, write it as ‘@exec_prefix@’.)
> >
> > Generally, $(exec_prefix) is used for directories that contain machine-specific files (such as executables and subroutine libraries), while $(prefix) is used directly for other directories.
> >
> > Running ‘make install’ with a different value of exec_prefix from the one used to build the program should not recompile the program.
> 
> I am not sure, but this could make live easier for package maintainers, given that they then can set PREFIX empty and EXEC_PREFIX=/usr. I mean, while I am at it... ;) Unless this goes to much out of scope for this PR, that is.

That's outside the scope of this PR. Please make a new one if you want to fix this

**Comment by WhitePeter** [comment]
> That's outside the scope of this PR. Please make a new one if you want to fix this

OK. So, now I think I have the final version, which also changes the default make target back to what I initially wanted: uninstall-keepconf. I reworked test/test-install.sh to account for dunstrc staying and did some cleanup in there.

**Comment by WhitePeter** [comment]
I guess that answers the question where the tests are taken from. ;)

If all is to your liking let me know, so I can finalize this by updating the README and changelog.

**Comment by WhitePeter** [comment]
Sorry, forgot to remove my debug settings from test-install.sh. This is it though. I won't touch anything until someone tells me otherwise.

**Review by fwSmit — APPROVED** [review]

**Comment by fwSmit** [comment]
This looks good now. Thank you!

**Comment by WhitePeter** [comment]
Thank you, too!

**Pull request merged**
